pkg/build/controller/controller.go

```diff
@@ -108,6 +163,44 @@ func (bc *BuildController) HandleBuild(build *buildapi.Build) error {
 	return nil
 }
 
+func (bc *BuildController) nextBuild(namespace, buildConfigName string) (*buildapi.Build, error) {
+	hasRunningBuilds := false
+	builds, err := buildutil.BuildConfigBuilds(bc.OSClient.Builds(namespace), buildConfigName, func(b buildapi.Build) bool {
+		// Detect if there are running builds
+		switch b.Status.Phase {
+		case buildapi.BuildPhaseRunning, buildapi.BuildPhasePending:
+			hasRunningBuilds = true
+			return false
+		}
+		// Only 'new' build can be scheduled to run next
+		return b.Status.Phase == buildapi.BuildPhaseNew
+	})
+	if hasRunningBuilds || err != nil {
+		return nil, err
+	}
+	var (
+		nextBuild           *buildapi.Build
+		previousBuildNumber int64
+	)
+	for i, b := range builds.Items {
+		buildNumber, err := buildutil.BuildNumber(&b)
+		if err != nil {
+			glog.Errorf("Error getting build number: %v", err)
+			continue
+		}
+		if previousBuildNumber == 0 {
+			nextBuild = &builds.Items[i]
+			previousBuildNumber = buildNumber
+			continue
+		}
```

Comment:
On Mon, Apr 18, 2016 at 6:40 AM, Michal Fojtik <notifications@github.com>
wrote:

> In pkg/build/controller/controller.go
> <https://github.com/openshift/origin/pull/8453#discussion_r60037443>:
>
> > +	}
> > +	var (
> > +		nextBuild           *buildapi.Build
> > +		previousBuildNumber int64
> > +	)
> > +	for i, b := range builds.Items {
> > +		buildNumber, err := buildutil.BuildNumber(&b)
> > +		if err != nil {
> > +			glog.Errorf("Error getting build number: %v", err)
> > +			continue
> > +		}
> > +		if previousBuildNumber == 0 {
> > +			nextBuild = &builds.Items[i]
> > +			previousBuildNumber = buildNumber
> > +			continue
> > +		}
>
> @smarterclayton <https://github.com/smarterclayton> the CI/CD executors
> might care about processing all created builds? This comes down whether we
> should kick a new build if there is already a build with the same inputs.
> The problem with that sort of validation is
> that the builder image state is not predictable (you might update the
> builder image between scheduled builds).
>

CI might - but not non-CI


> I agree that the "rapid" option, where every trigger will create a build
> that will get scheduled next and cancelling all
> the builds in the queue is valuable (and mostly like all users will use
> that for development). Maybe we need three
> scheduling policies:
>
>    - *parallel* - current behavior, every build gets executed immediately
>    after creation, no queue.
>    - *serial* - mimic the Jenkins build scheduling, builds are queued and
>    executed in the serial order
>    - *rapid* - (I struggle to give this a name ;-) - same as serial, but
>    without queue, the last build created is always executed as next,
>    cancelling the previous build.
>
>
"latest"


>
>
> @bparees <https://github.com/bparees> @liggitt
> <https://github.com/liggitt> what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453/files/68d0bdac5916b17673695e06451439bfa55941b9#r60037443>
>